### PR TITLE
WIP: noma/ocl integration

### PR DESCRIPTION
**Pull request by panau161**

### CODE REVIEW

File: make_ocl.sh [hunk 1]
```diff
@@ -40,12 +40,14 @@ NUM_ITERATIONS=26 # including warmup below
 NUM_WARMUP=1
 INTEL_PREFETCH_LEVEL=1 # sets -auto-prefetch-level= for OpenCL compilation
 
-HAM_PATH=thirdparty/ham
-CLU_PATH_INCLUDE=thirdparty/CLU.git
-CLU_PATH_LIB=thirdparty/CLU/clu_runtime
+HAM_PATH_INCLUDE=thirdparty/ham/include/
```
Comment (noma) [inline make_ocl.sh]: TODO: everything HAM can be replaced by thirdparty/noma/bmt, which is the updated stand-alone version of the benchmark timer, grep code for includes of the ham header, and adapt all codes

File: make_ocl.sh [hunk 2]
```diff
@@ -40,12 +40,14 @@ NUM_ITERATIONS=26 # including warmup below
...
+NOMAOCL_PATH_INCLUDE=thirdparty/ocl/include/
```
Comment (noma) [inline make_ocl.sh]: Please use `NOMA_<LIB_NAME>`, i.e. add `_`, e.g. `NOMA_OCL`

File: make_ocl.sh [hunk 3]
```diff
@@ -40,12 +40,14 @@ NUM_ITERATIONS=26 # including warmup below
...
+NOMAOCL_PATH_LIB=thirdparty/ocl/build/
 
-INCLUDE="-Iinclude -I${HAM_PATH}/include/ -I${OPENCL_INCLUDE_PATH} -I${CLU_PATH_INCLUDE}"
-LIB="-lrt -L${OPENCL_LIB_PATH} -l${OPENCL_LIB} ${CLU_PATH_LIB}/libclu_runtime.a"
+NOMAMISC_PATH_INCLUDE=thirdparty/misc/include/
+
+INCLUDE="-Iinclude -I${HAM_PATH_INCLUDE} -I${OPENCL_INCLUDE_PATH} -I${NOMAOCL_PATH_INCLUDE} -I${NOMAMISC_PATH_INCLUDE}"
+LIB="-lrt -lboost_program_options -L${OPENCL_LIB_PATH} -l${OPENCL_LIB} ${NOMAOCL_PATH_LIB}/libnoma_ocl.a ${NOMAOCL_PATH_LIB}/build.noma_typa/libnoma_typa.a"
```
Comment (noma) [inline make_ocl.sh]: also requires `-L${BOOST_ROOT}/lib` before `-lboost_program_options`

File: make_ocl.sh [hunk 4]
```diff
@@ -40,12 +40,14 @@ NUM_ITERATIONS=26 # including warmup below
...
-CLU_PATH_LIB=thirdparty/CLU/clu_runtime
+HAM_PATH_INCLUDE=thirdparty/ham/include/
+NOMAOCL_PATH_INCLUDE=thirdparty/ocl/include/
+NOMAOCL_PATH_LIB=thirdparty/ocl/build/
 
-INCLUDE="-Iinclude -I${HAM_PATH}/include/ -I${OPENCL_INCLUDE_PATH} -I${CLU_PATH_INCLUDE}"
-LIB="-lrt -L${OPENCL_LIB_PATH} -l${OPENCL_LIB} ${CLU_PATH_LIB}/libclu_runtime.a"
+NOMAMISC_PATH_INCLUDE=thirdparty/misc/include/
+
+INCLUDE="-Iinclude -I${HAM_PATH_INCLUDE} -I${OPENCL_INCLUDE_PATH} -I${NOMAOCL_PATH_INCLUDE} -I${NOMAMISC_PATH_INCLUDE}"
```
Comment (noma) [inline make_ocl.sh]: also requires `-I${BOOST_ROOT}/include`

File: plot_results.r [hunk 5]
```diff
@@ -38,8 +38,14 @@ if (!file.exists(file)) {
 	quit()
 }
 
+# get file line-by-line
```
Comment (noma) [inline plot_results.r:41]: Why is that necessary? 
K, got it, there is a pre-process step involved that filters the log using grep in a separate script, to keep file format and of exe and plot separated (the plot stuff should be relatively independent of the specific project, there's a newer version somewhere, should start a separate repo for that. ;-))

File: src/benchmark_ocl.cpp [hunk 6]
```diff
@@ -144,125 +119,106 @@ int main(void)
...
-	ocl_error_handler(err, "clCreateBuffer(hamiltonian_ocl)");
-	cl_mem sigma_in_ocl = clCreateBuffer(CLU_CONTEXT, CL_MEM_READ_WRITE, size_sigma_byte, 0, &err);
-	ocl_error_handler(err, "clCreateBuffer(sigma_in_ocl)");
-	cl_mem sigma_out_ocl = clCreateBuffer(CLU_CONTEXT, CL_MEM_READ_WRITE, size_sigma_byte, 0, &err);
-	ocl_error_handler(err, "clCreateBuffer(sigma_out_ocl)");
+	ocl_helper.write_device_info(std::cerr);
+
+	// allocate OpenCL device memory // DONE: replace with create_buffer() from noma::ocl, use C++ interface
+	noma::ocl::buffer hamiltonian_ocl = ocl_helper.create_buffer(CL_MEM_READ_ONLY,  size_hamiltonian_byte);
+	noma::ocl::buffer sigma_in_ocl    = ocl_helper.create_buffer(CL_MEM_READ_WRITE, size_sigma_byte      );
```
Comment (noma) [inline src/benchmark_ocl.cpp]: Not a big fan of this aligned style, as there are situations with very long lines with additional line breaks etc. where it does not work well, and then it cannot be applied as a general rule etc... I'm OK with it when it helps with readability a lot, but please remove at least the trailing spaces after the last argument

File: src/benchmark_ocl.cpp [hunk 7]
```diff
@@ -144,125 +119,106 @@ int main(void)
...
-			std::cerr << "OpenCL build log for: " << file_name << std::endl
-			          << cluGetBuildErrors(prog) << std::endl
-			          << "--- end of build log ---" << std::endl;
-			exit(-1);
-		}
-		cl_kernel kernel = clCreateKernel(prog, kernel_name.c_str(), &err);
-		ocl_error_handler(err, "clCreateKernel()");
+		// get kernel from programm using C++ OCL API
+		cl::Kernel kernel(prog, kernel_name.c_str(), &err);
+		noma::ocl::error_handler(err, "Error creating kernel.");
```
Comment (noma) [inline src/benchmark_ocl.cpp]: Let's add the kernel name for easier debugging:
`noma::ocl::error_handler(err, "Error creating kernel: '" + kernel_name + "'.");`

File: src/benchmark_ocl.cpp [hunk 8]
```diff
@@ -144,125 +119,106 @@ int main(void)
...
 
 		// set kernel arguments
-		err = clSetKernelArg(kernel, 0, sizeof(cl_mem), static_cast<const void*>(&sigma_in_ocl));
-		ocl_error_handler(err, "clSetKernelArg(0)");
-		err = clSetKernelArg(kernel, 1, sizeof(cl_mem), static_cast<const void*>(&sigma_out_ocl));
-		ocl_error_handler(err, "clSetKernelArg(1)");
-		err = clSetKernelArg(kernel, 2, sizeof(cl_mem), static_cast<const void*>(&hamiltonian_ocl));
-		ocl_error_handler(err, "clSetKernelArg(2)");
+		err = kernel.setArg(0, static_cast<cl::Buffer>(sigma_in_ocl));
```
Comment (noma) [inline src/benchmark_ocl.cpp:122]: The problem was here, we use the **typed** C++ API here, instead of throwing around void pointers and byte sizes, like the C-API calls did.

File: src/benchmark_ocl.cpp [hunk 9]
```diff
@@ -144,125 +119,106 @@ int main(void)
...
+		noma::ocl::error_handler(err, "kernel.setArg(0)");
+		err = kernel.setArg(1, static_cast<cl::Buffer>(sigma_out_ocl));
+		noma::ocl::error_handler(err, "kernel.setArg(1)");
+		err = kernel.setArg(2, static_cast<cl::Buffer>(hamiltonian_ocl));
+		noma::ocl::error_handler(err, "kernel.setArg(2)");
 		// GCC bug work-around (#ifdef __GNUC__) & type conversion
 		// http://stackoverflow.com/questions/19616610/c11-lambda-doesnt-take-const-variable-by-reference-why
+		// TODO: Is it okay to remove const keyword? Reference:
+		// Internet:                                    cl_int setArg(cl_uint index, size_type size, const void *argPtr)
+		// thirdparty/ocl/thirdparty/include/CL/cl.hpp: cl_int setArg(cl_uint index,  ::size_t size,       void* argPtr)
```
Comment (noma) [inline src/benchmark_ocl.cpp]: NOTE: We keep the tmp variables of defined-length-types here, as we have to make sure, the byte size still matches the OpenCL kernel signature types.

File: src/benchmark_ocl.cpp [hunk 10]
```diff
@@ -144,125 +119,106 @@ int main(void)
...
-	//     size_t    local[3];
-	//     size_t    offset[3];
-	// } clu_nd_range;
+	//     cl::NDRange offset;
+	//     cl::NDRange global;
+	//     cl::NDRange local;
+	// };
 	auto benchmark = [&](const std::string& file_name, const std::string& kernel_name,
-	                     const std::string& compile_options, size_t vec_length, clu_nd_range range,
+	                     const std::string& compile_options, size_t vec_length, noma::ocl::nd_range& range,
```
Comment (noma) [inline src/benchmark_ocl.cpp]: nd_range can be const here

File: src/benchmark_ocl.cpp [hunk 11]
```diff
@@ -41,42 +41,17 @@ using namespace ham::util;
 	#define WARP_SIZE 32
 #endif
 
-bool ocl_error_handler(cl_int err, const std::string& function_name, bool terminate = true)
+// integrate into benchmark()?
```
Comment (noma) [inline src/benchmark_ocl.cpp]: Yes, it's the only call-site, so let's just remove that function and put its body directly into the benchmark lambda.

File: src/benchmark_ocl.cpp [hunk 12]
```diff
@@ -144,125 +119,106 @@ int main(void)
...
+	//     cl::NDRange local;
+	// };
 	auto benchmark = [&](const std::string& file_name, const std::string& kernel_name,
-	                     const std::string& compile_options, size_t vec_length, clu_nd_range range,
+	                     const std::string& compile_options, size_t vec_length, noma::ocl::nd_range& range,
 	                     decltype(&transform_matrices_aos_to_aosoa) transformation_sigma,
 	                     bool scale_hamiltonian,
-	                     decltype(&transform_matrix_aos_to_soa) transformation_hamiltonian)
+	                     decltype(&transform_matrix_aos_to_soa) transformation_hamiltonian,
+	                     noma::ocl::helper& ocl_helper)
```
Comment (noma) [inline src/benchmark_ocl.cpp]: we can remove `ocl_helper` from the argument list, as we capture all variables of `benchmark`'s surrounding code by reference `[&]` (Google C++ Lambda or check cppreference.com for further input). We only need arguments that differ during calls.

File: thirdparty/build.sh [hunk 13]
```diff
@@ -5,13 +5,14 @@
...
-# build CLU
-mkdir -p CLU
-cd CLU
-cmake -DCMAKE_CXX_COMPILER=g++ -DCMAKE_C_COMPILER=gcc ../CLU.git
+# noma-ocl
+# Standalone/Library Build
+cd ocl
+mkdir -p build
+cd build
+cmake -DNOMA_OCL_STANDALONE=TRUE
```
Comment (noma) [inline thirdparty/build.sh]: `..` at the end needed, also the build type `Release` should be set.

**Review by noma — CHANGES_REQUESTED** [review]

File: make_ocl.sh (hunk 2 above)
Comment (panau161) [reply make_ocl.sh]: Done with 83f8f7073908672448c504abe1470de885dc9740.

File: make_ocl.sh (hunk 4 above)
Comment (panau161) [reply make_ocl.sh]: Done with 83f8f7073908672448c504abe1470de885dc9740.

File: make_ocl.sh (hunk 3 above)
Comment (panau161) [reply make_ocl.sh]: Done with 83f8f7073908672448c504abe1470de885dc9740.

File: src/benchmark_ocl.cpp (hunk 11 above)
Comment (panau161) [reply src/benchmark_ocl.cpp]: Done with a758cba0fa201d64c477ff298ffa7ea784117a09.

File: src/benchmark_ocl.cpp (hunk 6 above)
Comment (panau161) [reply src/benchmark_ocl.cpp]: Changed alignment style where it is not necessary with a758cba0fa201d64c477ff298ffa7ea784117a09.

File: src/benchmark_ocl.cpp (hunk 7 above)
Comment (panau161) [reply src/benchmark_ocl.cpp]: Done with a758cba0fa201d64c477ff298ffa7ea784117a09.

File: src/benchmark_ocl.cpp (hunk 10 above)
Comment (panau161) [reply src/benchmark_ocl.cpp]: Done with a758cba0fa201d64c477ff298ffa7ea784117a09.

File: thirdparty/build.sh (hunk 13 above)
Comment (panau161) [reply thirdparty/build.sh]: Done with 1b6a3fa8e4ecbdd8c781c3b0e5ee0673a482e07b.

File: src/benchmark_ocl.cpp (hunk 12 above)
Comment (panau161) [reply src/benchmark_ocl.cpp]: Done with fa04e8c75f867e6b7b543462a86dc012327d7fd3.

File: cmake_ocl.sh [hunk 14]
```diff
@@ -31,9 +31,14 @@ build()
 	mkdir ${DIR_NAME}
 	cd ${DIR_NAME}
 
-	cmake -DCMAKE_BUILD_TYPE=Release ${BUILD_OPTIONS} ${CMAKE_OPTIONS} ..
+	CXX=`which CC` CC=`which cc` cmake -DCMAKE_BUILD_TYPE=Release ${BUILD_OPTIONS} ${CMAKE_OPTIONS} ..
```
Comment (noma) [inline cmake_ocl.sh]: This needs some generic solution... 
We could remove the variables before `cmake` and call the build script like this
```
CXX=`which CC` CC=`which cc` cmake_ocl.sh # etc..
```

File: cmake_ocl.sh [hunk 15]
```diff
@@ -43,6 +48,7 @@ usage ()
 	echo -e "\t-k\t Build KNL variant.";
 	echo -e "\t-a\t Build KNC accelerator variant.";
 	echo -e "\t-g\t Build GPU variant.";
+	echo -e "\t-d\t Enable debug mode.";
```
Comment (noma) [inline cmake_ocl.sh]: Maybe call it enable verbose, non-parallel build, as we are still doing a release build without debug symbols.
```
Enable verbose make for debugging the build process.
```
